fix: Must not be able to start Job Card if it is related to Work Order that is not started yet

FHenry opened the pull request:
related to #29071

FHenry committed:
fix: Cannot start Job strat if related to Work Order not started yet

FHenry committed:
fix: Cannot start Job strat if related to Work Order not started yet

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into dev_29071

FHenry committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into dev_29071

FHenry committed:
Merge branch 'dev_29071' of github.com:FHenry/erpnext into dev_29071

FHenry committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into dev_29071

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into dev_29071

FHenry committed:
test

FHenry committed:
test

FHenry committed:
Merge branch 'dev_29071' of github.com:FHenry/erpnext into dev_29071

FHenry committed:
fix siders

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'frappe:develop' into dev_29071

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry commented:
Hi @marination Do you need more explanation from me about this PR ?

FHenry committed:
Merge branch 'develop' into dev_29071

FHenry committed:
Merge branch 'develop' into dev_29071

johnlongland commented:
It will be great if this can be included into the core. Any idea when this might happen ?

marination commented on erpnext/manufacturing/doctype/job_card/job_card.js:
```diff
@@ -73,7 +73,24 @@ frappe.ui.form.on('Job Card', {
+			//if Job Card is link to Work Order, the job card must not be able to start if Work Order not "Started"
+			// and if stock mvt for WIP is required
+			if (frm.doc.work_order) {
+				frappe.db.get_value('Work Order', frm.doc.work_order, 'skip_transfer').then((r) => {
```
Lets get whatever we need in one db call. `result` will have to be accessed as per whatever is returned (`r.message` wont be required)
```suggestion
				frappe.db.get_value('Work Order', frm.doc.work_order, ['skip_transfer', 'status']).then((result) => {
```

marination commented on erpnext/manufacturing/doctype/job_card/job_card.js:
```diff
@@ -73,7 +73,24 @@ frappe.ui.form.on('Job Card', {
+			// and if stock mvt for WIP is required
+			if (frm.doc.work_order) {
+				frappe.db.get_value('Work Order', frm.doc.work_order, 'skip_transfer').then((r) => {
+					if (r.message.skip_transfer !== 1 ) {
```
The last else can be eliminated by combining both if conditions.
Also does "Completed" work order need a timer button?
```suggestion
					if (result.skip_transfer !== 1  && result.status == 'In Process') {
					     frm.trigger("prepare_timer_buttons");
					}
```

FHenry committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into dev_29071

FHenry committed:
PR review

FHenry committed:
Merge branch 'dev_29071' of github.com:FHenry/erpnext into dev_29071

FHenry commented:
@marination I've follow your instruction, and it's more readable like that. But result is not enought, we need result.message, like is it done line 95. It seems that frappe.db.get_value return result as r.message

marination committed:
chore: Code cleanup

- Better short circuit for if condition (make it such that both conditions dont always have to be computed)
- Remove `r.message` extraction by avoiding `then()`

marination committed:
chore: Remove unnecessary json change

FHenry commented:
@Mergifyio backport version-13-hotfix
